### PR TITLE
🌟 alicloud: effective permissions and public-bucket verdicts

### PR DESCRIPTION
**Pull request by tas50**
## Why

`#9641` typed the RAM identity graph, so `alicloud.ram.user.attachedPolicies` reaches a real `alicloud.ram.policy`. But the graph then stopped at a string: `policyDocument` returns the document as raw JSON. Every question worth asking of a permission policy ("who holds admin", "which custom policy allows everything", "which grants are unscoped") meant parsing JSON inside MQL.

Two other gaps sat next to it:

- `attachedPolicies` is **direct attachments only**. Whenever permissions are granted through a group, it understates what a user can actually do, which is exactly where over-privilege hides.
- `alicloud.oss.bucket` exposes `acl`, `policy` and `blockPublicAccess` separately, so answering "is this bucket reachable by anonymous callers" required reimplementing the precedence between them per query.

## What

**`alicloud.ram.policy.statement`** breaks the default policy version into rows carrying `effect`, `action`, `notAction`, `resource`, `notResource` and `condition`. It clears the sub-resource bar on complexity rather than identity, so its `__id` is synthetic and stays hidden.

```
alicloud.ram.policies.where(policyType == "Custom") {
  policyName
  statements.where(effect == "Allow") { action resource }
}
```

**Three derived verdicts on `alicloud.ram.policy`:**

| Field | True when |
|---|---|
| `allowsAdminAccess` | an Allow statement grants action `*` on resource `*` (or `acs:*:*:*:*`) with no `NotAction`/`NotResource` exception |
| `hasWildcardAction` | any Allow statement names an action containing `*`, including prefix forms like `ecs:Describe*` |
| `hasUnscopedResource` | any Allow statement applies to a whole service rather than named resources |

**`alicloud.ram.user.effectivePolicies`** unions direct attachments with the attachments of every group the user belongs to, deduplicated on the `(policyType, policyName)` pair:

```
alicloud.ram.users.where(effectivePolicies.any(allowsAdminAccess))
```

**`alicloud.oss.bucket.isPublic`** folds a public-read canned ACL and a bucket policy naming a `*` principal into one verdict, with block public access overriding both.

## Two decisions worth flagging

**`hasUnscopedResource` reads only the relative ID.** A first cut defined the resource predicate as "the name contains a `*`" and a unit test immediately caught that this fires on almost every policy in existence: the region and account fields of an ACS name are wildcarded in normal usage. `acs:oss:*:*:mybucket/*` names one bucket; `acs:oss:*:*:*` names all of OSS. Only the fifth field decides, and the name says `Unscoped` rather than `WildcardResource` so it can't be misread.

**A `Condition` does not clear `allowsAdminAccess`.** A conditioned grant is still unrestricted wherever the condition holds, so it counts. `NotAction`/`NotResource` do disqualify a statement, because they carve a genuine exception out of the grant.

## Cost

`GetPolicy` was already returning `DefaultPolicyVersion.PolicyDocument` in `initAlicloudRamPolicy` and discarding it. It is now seeded onto the resource, so `statements` and the three verdicts cost **no extra call** on any policy reached through an attachment, and one memoized call otherwise. The four derived fields share a single parse.

## Tests

`policy_test.go` covers the parser against the shape variance the grammar allows, since all three document flavors (RAM permission policy, role trust policy, OSS bucket policy) share one envelope:

- `Statement` as a lone object or an array
- every list-valued member as a bare string or an array
- `Principal` in all three forms: keyed object, bare array, lone string
- the `resourceIsUnscoped` boundary cases, the `Deny`-never-grants rule, and case-insensitive `Effect`

## Verification

`gofmt`, `go vet`, `go build ./...` and `go test ./resources/` all pass. **Not yet live-verified** against a real Alibaba Cloud account, consistent with the `#9641`-`#9647` stack, which is still owed a batched verification run.

## Follow-up left out of scope

`alicloud.ram.role.assumeRolePolicyDocument` is still a raw string. The same parser already handles trust-policy `Principal` shapes, so `trustedPrincipals` and a `hasWildcardPrincipal` verdict on the role are a small next step.

### CODE REVIEW

File: providers/alicloud/resources/ram.go [hunk 1]
```diff
@@ -1086,26 +1155,117 @@ func resolveRamPolicy(runtime *plugin.Runtime, policyName, policyType *string) (
...
+	statementsErr   error
+}
+
+// fetchDocument returns the policy document of the default version, calling
+// GetPolicy at most once per policy across every field that needs it.
+func (r *mqlAlicloudRamPolicy) fetchDocument() (string, error) {
+	r.documentOnce.Do(func() {
+		if r.cacheDocument != "" {
+			return
+		}
```
[the remark this answers is not in the document]
Comment (tas50) [reply providers/alicloud/resources/ram.go:1177]: Fixed in #9695 with a `documentSeeded` flag. Went one step past the suggestion: init sets the flag even when the response has no `DefaultPolicyVersion` at all, since a repeat call returns the same nothing, so that path was paying for a redundant round-trip too.

File: providers/alicloud/resources/oss.go [hunk 2]
```diff
@@ -454,3 +454,39 @@ func (a *mqlAlicloudOssBucket) blockPublicAccess() (bool, error) {
...
+	policy := a.GetPolicy()
+	if policy.Error != nil {
+		return false, policy.Error
+	}
+	statements, err := parsePolicyDocument(policy.Data)
+	if err != nil {
+		// an unparseable policy is not evidence of exposure, and the raw
+		// document stays available through the policy field
+		return false, nil
+	}
```
[the remark this answers is not in the document]
Comment (tas50) [reply providers/alicloud/resources/oss.go:490]: Added in #9695. Used `log.Warn` rather than the provider's usual `Debug` for tolerated failures, because this one degrades a security verdict rather than dropping an optional detail. `a.__id` isn't reachable from Go, so the log uses `a.Name.Data`, which is the same value.

File: providers/alicloud/resources/policy.go [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+// resourceIsUnscoped reports whether an ACS resource name selects everything
+// its service offers rather than a named resource.
+//
+// Only the relative-id, the fifth field of
+// acs:<service>:<region>:<account>:<relative-id>, is examined. The region and
+// account fields are wildcarded in most real policies, so treating any "*" in
+// the name as unscoped would flag nearly every statement ever written:
+// "acs:oss:*:*:mybucket/*" names one bucket, while "acs:oss:*:*:*" names all of
+// OSS.
+func resourceIsUnscoped(resource string) bool {
```
[the remark this answers is not in the document]
Comment (tas50) [reply providers/alicloud/resources/policy.go:230]: Comment added in #9695.

Writing it turned up a real gap: the test named "a relative-id containing colons is kept whole" used `acs:ram::123:user/alice`, which has **no colon in its relative-id** — it was testing the empty-region case under the wrong name. Both are now covered separately, and the colon case uses inputs that would actually regress if the limit changed:

```go
assert.False(t, resourceIsUnscoped("acs:ram::123:role/admin:extra"))
assert.False(t, resourceIsUnscoped("acs:oss:*:*:bucket/prefix:*"))
```

Without the limit the second one splits to a trailing `*` and reads as a whole-service grant, i.e. a false positive on `hasUnscopedResource`.
